### ISSUE TITLE
Ceph OSD fails to init after node reboot

### ISSUE THREAD

**Issue by lix0st** [rook/rook#1754]
Is this a bug report or feature request?

* Bug Report

**Bug Report**

What happened:

After a node reboot (drained the node beforehand) the OSD pod on that node is creating new OSD.
In the OSD pod there are errors concerning an OSD that can't be initialized (ERROR: osd init failed: (2) No such file or directory) :
```
kubectl -n rook-ceph logs rook-ceph-osd-phk8s-node12-l7dsp -f

2018-05-30 11:27:57.527317 I | proc: starting process [ceph-osd --foreground --id=3 --cluster=rook-ceph --osd-data=/var/lib/rook/osd3 --conf=/var/lib/rook/osd3/rook-ceph.config --keyring=/var/lib/rook/osd3/keyring --osd-uuid=16ff655b-7072-43e5-832a-5e95a5e27e70] again after 30.0 seconds
2018-05-30 11:28:15.595593 I | exec: Running command: ceph osd dump --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/107621560
2018-05-30 11:28:16.179866 I | cephosd: osd.3 is marked 'DOWN'
2018-05-30 11:28:27.527494 I | exec: Running command: ceph-osd --foreground --id=3 --cluster=rook-ceph --osd-data=/var/lib/rook/osd3 --conf=/var/lib/rook/osd3/rook-ceph.config --keyring=/var/lib/rook/osd3/keyring --osd-uuid=16ff655b-7072-43e5-832a-5e95a5e27e70
2018-05-30 11:28:27.528058 I | proc: retry (total 6). started process [ceph-osd --foreground --id=3 --cluster=rook-ceph --osd-data=/var/lib/rook/osd3 --conf=/var/lib/rook/osd3/rook-ceph.config --keyring=/var/lib/rook/osd3/keyring --osd-uuid=16ff655b-7072-43e5-832a-5e95a5e27e70]
2018-05-30 11:28:27.566609 I | osd3: 2018-05-30 11:28:27.566429 7f0d7e364d80  0 ceph version 12.2.5 (cad919881333ac92274171586c827e01f554a70a) luminous (stable), process (unknown), pid 1151
2018-05-30 11:28:27.566952 I | osd3: 2018-05-30 11:28:27.566765 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.566972 I | osd3: 2018-05-30 11:28:27.566904 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.566982 I | osd3: 2018-05-30 11:28:27.566936 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.567017 I | osd3: 2018-05-30 11:28:27.566967 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.568751 I | osd3: starting osd.3 at - osd_data /var/lib/rook/osd3 /var/lib/ceph/osd/rook-ceph-3/journal
2018-05-30 11:28:27.573087 I | osd3: 2018-05-30 11:28:27.572930 7f0d7e364d80  0 pidfile_write: ignore empty --pid-file
2018-05-30 11:28:27.640152 I | osd3: 2018-05-30 11:28:27.639957 7f0d7e364d80  0 load: jerasure load: lrc load: isa 
2018-05-30 11:28:27.640177 I | osd3: 2018-05-30 11:28:27.640058 7f0d7e364d80  1 bdev create path /var/lib/rook/osd3/block type kernel
2018-05-30 11:28:27.640188 I | osd3: 2018-05-30 11:28:27.640075 7f0d7e364d80  1 bdev(0x55acb637ac00 /var/lib/rook/osd3/block) open path /var/lib/rook/osd3/block
2018-05-30 11:28:27.640200 I | osd3: 2018-05-30 11:28:27.640093 7f0d7e364d80 -1 bdev(0x55acb637ac00 /var/lib/rook/osd3/block) open open got: (2) No such file or directory
2018-05-30 11:28:27.640846 I | osd3: 2018-05-30 11:28:27.640770 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.640916 I | osd3: 2018-05-30 11:28:27.640861 7f0d7e364d80  1 bdev create path /var/lib/rook/osd3/block type kernel
2018-05-30 11:28:27.640938 I | osd3: 2018-05-30 11:28:27.640873 7f0d7e364d80  1 bdev(0x55acb62a3600 /var/lib/rook/osd3/block) open path /var/lib/rook/osd3/block
2018-05-30 11:28:27.640955 I | osd3: 2018-05-30 11:28:27.640896 7f0d7e364d80 -1 bdev(0x55acb62a3600 /var/lib/rook/osd3/block) open open got: (2) No such file or directory
2018-05-30 11:28:27.640972 I | osd3: 2018-05-30 11:28:27.640921 7f0d7e364d80 -1 osd.3 0 OSD:init: unable to mount object store
2018-05-30 11:28:27.640989 I | osd3: 2018-05-30 11:28:27.640942 7f0d7e364d80 -1  ** ERROR: osd init failed: (2) No such file or directory
2018-05-30 11:28:27.645751 I | osd3: 2018-05-30 11:28:27.566765 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.645769 I | osd3: 2018-05-30 11:28:27.566904 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.645777 I | osd3: 2018-05-30 11:28:27.566936 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.645787 I | osd3: 2018-05-30 11:28:27.566967 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.645802 I | osd3: 2018-05-30 11:28:27.640093 7f0d7e364d80 -1 bdev(0x55acb637ac00 /var/lib/rook/osd3/block) open open got: (2) No such file or directory
2018-05-30 11:28:27.645818 I | osd3: 2018-05-30 11:28:27.640770 7f0d7e364d80 -1 bluestore(/var/lib/rook/osd3/block) _read_bdev_label failed to open /var/lib/rook/osd3/block: (2) No such file or directory
2018-05-30 11:28:27.645831 I | osd3: 2018-05-30 11:28:27.640896 7f0d7e364d80 -1 bdev(0x55acb62a3600 /var/lib/rook/osd3/block) open open got: (2) No such file or directory
2018-05-30 11:28:27.645843 I | osd3: 2018-05-30 11:28:27.640921 7f0d7e364d80 -1 osd.3 0 OSD:init: unable to mount object store
2018-05-30 11:28:27.645860 I | osd3: 2018-05-30 11:28:27.640942 7f0d7e364d80 -1  ** ERROR: osd init failed: (2) No such file or directory
2018-05-30 11:28:27.645917 I | proc: process 1151 completed.  Exited: true, ExitStatus: 1, Signaled: false, Signal: -1, &{Path:/usr/bin/ceph-osd Args:[ceph-osd --foreground --id=3 --cluster=rook-ceph --osd-data=/var/lib/rook/osd3 --conf=/var/lib/rook/osd3/rook-ceph.config --keyring=/var/lib/rook/osd3/keyring --osd-uuid=16ff655b-7072-43e5-832a-5e95a5e27e70] Env:[] Dir: Stdin:<nil> Stdout:0xc4202a6108 Stderr:0xc4202a6120 ExtraFiles:[] SysProcAttr:<nil> Process:0xc42038e450 ProcessState:<nil> ctx:<nil> lookPathErr:<nil> finished:false childFiles:[0xc4202a6128 0xc4202a6108 0xc4202a6120] closeAfterStart:[0xc4202a6108 0xc4202a6120 0xc4202a6128] closeAfterWait:[0xc4202a6100 0xc4202a6118] goroutine:[] errch:0xc4207568a0 waitDone:<nil>}
2018-05-30 11:28:27.645956 I | proc: starting process [ceph-osd --foreground --id=3 --cluster=rook-ceph --osd-data=/var/lib/rook/osd3 --conf=/var/lib/rook/osd3/rook-ceph.config --keyring=/var/lib/rook/osd3/keyring --osd-uuid=16ff655b-7072-43e5-832a-5e95a5e27e70] again after 30.0 seconds
```
The OSD pod created 3 new OSD's (even I have only configured 3). So now I end up having 3 additional OSD.
Two of the previous (before reboot) OSD are UP and one is down (because of the above error I suppose):
```
# ceph -s
  cluster:
    id:     227beec6-248a-4f48-8dff-5441de671d52
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum rook-ceph-mon0,rook-ceph-mon1,rook-ceph-mon2
    mgr: rook-ceph-mgr0(active)
    osd: 12 osds: 11 up, 11 in
 
  data:
    pools:   1 pools, 256 pgs
    objects: 0 objects, 0 bytes
    usage:   11397 MB used, 6958 GB / 6969 GB avail
    pgs:     256 active+clean

# ceph osd tree
ID CLASS WEIGHT  TYPE NAME             STATUS REWEIGHT PRI-AFF 
-1       7.25909 root default                                  
-2       1.81477     host phk8s-node11                         
 0   hdd 0.45329         osd.0             up  1.00000 1.00000 
 1   hdd 0.45329         osd.1             up  1.00000 1.00000 
 2   hdd 0.90819         osd.2             up  1.00000 1.00000 
-5       3.62955     host phk8s-node12                         
 3   hdd 0.45329         osd.3           down        0 1.00000 
 4   hdd 0.45329         osd.4             up  1.00000 1.00000 
 5   hdd 0.90819         osd.5             up  1.00000 1.00000 
 9   hdd 0.45329         osd.9             up  1.00000 1.00000 
10   hdd 0.45329         osd.10            up  1.00000 1.00000 
11   hdd 0.90819         osd.11            up  1.00000 1.00000 
-7       1.81477     host phk8s-node13                         
 6   hdd 0.45329         osd.6             up  1.00000 1.00000 
 7   hdd 0.45329         osd.7             up  1.00000 1.00000 
 8   hdd 0.90819         osd.8             up  1.00000 1.00000 
```

I don't really know how it is possible to have 3 additional OSD's. So I end up having 6 OSD on the node that hase only 3 HDD (and only 3 OSD configured)
Did I used a false procedure to reboot the node?

What you expected to happen:
After a single node reboot the CEPH cluster should be healthy without additional OSD's.

How to reproduce it (minimal and precise):
I've deployed the rook operator using the helm chart and then deployed a CEPH cluster with the following YAML:
```
apiVersion: ceph.rook.io/v1alpha1
kind: Cluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  dataDirHostPath: /var/lib/rook
  # cluster level storage configuration and selection
  network:
    hostNetwork: true
  monCount: 3
  storage:
    useAllNodes: false
    useAllDevices: false
    deviceFilter:
    location:
    config:
      metadataDevice:
      databaseSizeMB: "1024" # this value can be removed for environments with normal sized disks (100 GB or larger)
      journalSizeMB: "1024"  # this value can be removed for environments with normal sized disks (20 GB or larger)
    nodes:
    - name: "phk8s-node11"
      devices:
      - name: "sdb"
      - name: "sdc"
      - name: "sdd"
      config:
        storeType: bluestore
    - name: "phk8s-node12"
      devices:
      - name: "sdb"
      - name: "sdc"
      - name: "sdd"
      config:
        storeType: bluestore
    - name: "phk8s-node13"
      devices:
      - name: "sdb"
      - name: "sdc"
      - name: "sdd"
      config:
        storeType: bluestore

```

All went fine and the cluster is UP and healthy:
```
kubectl -n rook-ceph get pods
NAME                               READY     STATUS    RESTARTS   AGE
rook-ceph-mgr0-7c9c597977-rktlc    1/1       Running   0          3m
rook-ceph-mon0-c2sbw               1/1       Running   0          4m
rook-ceph-mon1-l5j7q               1/1       Running   0          4m
rook-ceph-mon2-hbclk               1/1       Running   0          4m
rook-ceph-osd-phk8s-node11-d75kb   1/1       Running   0          3m
rook-ceph-osd-phk8s-node12-zgg9n   1/1       Running   0          2m
rook-ceph-osd-phk8s-node13-xp4hd   1/1       Running   0          1m
rook-ceph-tools                    1/1       Running   0          2m



[root@rook-ceph-tools /]# ceph -s
  cluster:
    id:     227beec6-248a-4f48-8dff-5441de671d52
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum rook-ceph-mon0,rook-ceph-mon1,rook-ceph-mon2
    mgr: rook-ceph-mgr0(active)
    osd: 9 osds: 9 up, 9 in
 
  data:
    pools:   0 pools, 0 pgs
    objects: 0 objects, 0 bytes
    usage:   9265 MB used, 5565 GB / 5575 GB avail
    pgs: 

ceph osd tree
ID CLASS WEIGHT  TYPE NAME             STATUS REWEIGHT PRI-AFF 
-1       7.25909 root default                                  
-2       1.81477     host phk8s-node11                         
 0   hdd 0.45329         osd.0             up  1.00000 1.00000 
 1   hdd 0.45329         osd.1             up  1.00000 1.00000 
 2   hdd 0.90819         osd.2             up  1.00000 1.00000 
-5       3.62955     host phk8s-node12                         
 3   hdd 0.45329         osd.3             up  1.00000 1.00000  
 4   hdd 0.45329         osd.4             up  1.00000 1.00000 
 5   hdd 0.90819         osd.5             up  1.00000 1.00000 
-7       1.81477     host phk8s-node13                         
 6   hdd 0.45329         osd.6             up  1.00000 1.00000 
 7   hdd 0.45329         osd.7             up  1.00000 1.00000 
 8   hdd 0.90819         osd.8             up  1.00000 1.00000 

```

Then I reboot one node (phk8s-node12) by draining all pods:
```
kubectl drain phk8s-node12 --ignore-daemonsets --delete-local-data

#Then reboot phk8s-node12
```
After the reboot I uncordon the node to schedule the pods on it (kubectl uncorn phk8s-node12).
And now I end up having 3 more OSD's

**Environment**:
* OS (e.g. from /etc/os-release): `Ubuntu 16.04.4 LTS`
* Kernel (e.g. `uname -a`): `4.4.0-127-generic`
* Cloud provider or hardware configuration: 
```
Bare metal:
3 vms as k8s master
5 Physical server as k8s workers (3 servers for ceph with each 3 HDD - 2x500Gb 1x1TB) 
```
* Rook version (use `rook version` inside of a Rook Pod): `rook: v0.7.0-158.gb2a3579`
* Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-21T09:17:39Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-21T09:05:37Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

```
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Bare metal
* Ceph status (use `ceph health` in the [Rook toolbox](https://Rook.io/docs/Rook/master/toolbox.html)): See above

**Comment by benceszikora** [comment]
I have the same issue (on a bit older version v0.7.0-134.g9b5c092, kubernetes 1.9.3, RHEL 7.3). I was just testing restarting the OSD pod and I am getting the same error.

**Comment by lix0st** [comment]
@benceszikora Sorry I don't know. I've started using rook two weeks ago by using the `master`. I'm also trying to get a production cluster, but as you say, for now it's not possible with this bug.
@galexrt Thanks for the info, I will try `v0.7.1`

**Comment by lix0st** [comment]
I've just tested with `v0.7.1` and I can confirm that this bug isn't happening. I can reboot nodes without any new OSDs appearing and having a healthy cluster.

**Comment by benceszikora** [comment]
Thanks, `v0.7.1-4.g4233e03` seems to work fine for me as well.

**Comment by travisn** [comment]
I noticed this on Friday as well and have a fix with #1762. It is related to the same change as #1748, though that symptom actually isn't the cause for this issue. I'll get the fix in today, just need to add a test for it.

**Comment by devthane** [comment]
I had this same issue happen recently on 0.9, but I'm not sure what caused it, maybe my own mistake?

Anyways just recreating the cluster wasn't an option for me (that I liked).
Instead deleted the OSD and MON deployments that were set to run on the affected node.
I then removed the rook related directory (/var/lib/rook) and I also ran wipefs -a on the drive that was affected.
Finally I deleted the operator pod and let it recreate everything afresh on the node.
Everything seems to be starting up fine now, though obviously I lost everything on the problem osd.

I'll update if the problem re-emerges on my next reboot.
